src/RemediationRoleSaver.cpp

```diff
@@ -79,35 +151,195 @@ void RemediationSaverBase::saveToFile(const QString& filename)
-    mFiletypeTemplate = QObject::tr("ansible playbook (*.%1)");
-    mFixTemplate = QObject::tr("urn:xccdf:fix:script:ansible");
+    mArfFile.setAutoRemove(true);
+    mArfFile.open();
+    mArfFile.write(arfContents);
+    mArfFile.close();
 }
 
 
-PuppetRemediationSaver::PuppetRemediationSaver(QWidget* parentWindow, ScanningSession* session):RemediationSaverBase(parentWindow, session)
+void ResultBasedLibraryRemediationSaver::saveToFile(const QString& filename)
 {
-    mSaveMessage = QObject::tr("Save remediation role as a puppet manifest");
-    mFiletypeExtension = "pp";
-    mFiletypeTemplate = QObject::tr("puppet manifest (*.%1)");
-    mFixTemplate = QObject::tr("urn:xccdf:fix:script:puppet");
+    struct oscap_source* source = oscap_source_new_from_file(mArfFile.fileName().toUtf8().constData());
+    oscap_document_type_t document_type = oscap_source_get_scap_type(source);
+    if (document_type != OSCAP_DOCUMENT_ARF) {
+        throw std::runtime_error("Expected an ARF file");
+    }
+
+    struct ds_rds_session* arf_session = ds_rds_session_new_from_source(source);
+    if (arf_session == NULL) {
+        throw std::runtime_error("Couldn't open ARF session");
+    }
+    struct oscap_source *report_source = ds_rds_session_select_report(arf_session, NULL);
+    if (report_source == NULL) {
+        throw std::runtime_error("Couldn't get report source from the ARF session");
+    }
+    struct oscap_source *report_request_source = ds_rds_session_select_report_request(arf_session, NULL);
+    if (report_request_source == NULL) {
+        throw std::runtime_error("Couldn't get report request source from the ARF session");
+    }
+
+    struct xccdf_session* session = xccdf_session_new_from_source(oscap_source_clone(report_request_source));
+    if (xccdf_session_add_report_from_source(session, oscap_source_clone(report_source))) {
+        throw std::runtime_error("Couldn't get report request source from the ARF session");
+    }
+    oscap_source_free(source);
+
+    if (session == NULL)
+        throw std::runtime_error("Couldn't get XCCDF session from the report source");
+
+    xccdf_session_set_loading_flags(session, XCCDF_SESSION_LOAD_XCCDF);
+    if (xccdf_session_load(session) != 0)
+        throw std::runtime_error("Couldn't load XCCDF");
+
+    // "" for profile should work, as it works when passed to the command-line oscap
+    if (xccdf_session_build_policy_from_testresult(session, "") != 0)
+        throw std::runtime_error("Couldn't get build policy from testresult");
+
+    struct xccdf_policy* policy = xccdf_session_get_xccdf_policy(session);
+    struct xccdf_result* result = xccdf_policy_get_result_by_id(policy, xccdf_session_get_result_id(session));
+    /* Result-oriented fixes */
+
+    QFile outputFile(filename);
+    outputFile.open(QIODevice::WriteOnly);
+
+    const int rc = xccdf_policy_generate_fix(policy, result, mTemplateString.toUtf8().constData(), outputFile.handle());
+    if (!outputFile.flush())
```

Comment:
doesn't this flush the Qt cache? Is this defined in the docs? We are using the handle directly and AFAIK Qt guarantees that the file will exist when we call "handle()", however I don't think flush() behavior is well defined for this case.

Reply:
[The docs}(http://doc.qt.io/qt-4.8/qfile.html#close) tell that `close()` calls `flush()`, but ignores any errors. I have verified that if there is nothing to write, the `xccdf_policy_generate_fix` doesn't complain, and `close()` doesn't indicate anything either. The only way how to verify that everything went OK seems to be to call `flush` manually, check for errors and then proceed with `close()`.